sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+  }
+
+  return kDifUsbdevInPacketWriteBytesResultOK;
+}
+
+dif_usbdev_tx_packet_send_result_t dif_usbdev_tx_packet_send(
+    dif_usbdev_t *usbdev, uint32_t endpoint) {
+  if (usbdev == NULL) {
+    return kDifUsbdevInPacketSendResultInvalidArgument;
+  }
+
+  if (!is_valid_endpoint(endpoint)) {
+    return kDifUsbdevInPacketSendResultGeneralError;
+  }
+
+  // Make sure that there is an active write operation for the endpoint
+  if (!buffer_op_is_writing(usbdev, endpoint)) {
+    return kDifUsbdevInPacketSendResultNotWriting;
+  }
+
+  // Stop writing to buffer for this endpoint
+  buffer_op_stop_write(usbdev, endpoint);
+  // Mark the packet as ready for transmission
+  mmio_region_nonatomic_set_bit32(
+      usbdev->base_addr, kEndpointHwInfos[endpoint].config_in_reg_offset,
+      USBDEV_CONFIGIN0_RDY0);
+
+  return kDifUsbdevInPacketSendResultOK;
+}
+
+dif_usbdev_tx_packet_get_status_result_t dif_usbdev_tx_packet_get_status(
+    dif_usbdev_t *usbdev, uint32_t endpoint,
+    dif_usbdev_tx_packet_status_t *status) {
+  if (usbdev == NULL || status == NULL) {
+    return kDifUsbdevTxPacketGetStatusResultInvalidArgument;
+  }
+
+  if (!is_valid_endpoint(endpoint)) {
+    return kDifUsbdevTxPacketGetStatusResultGeneralError;
+  }
+
+  // Make sure that there is no active write buffer operation for the endpoint
+  if (buffer_op_is_writing(usbdev, endpoint)) {
+    *status = kDifUsbdevTxPacketStatusStillWriting;
+    return kDifUsbdevTxPacketGetStatusResultOK;
+  }
+
+  // HW details
+  const uint32_t config_in_reg_offset =
+      kEndpointHwInfos[endpoint].config_in_reg_offset;
+  const uint32_t endpoint_bit_index = kEndpointHwInfos[endpoint].bit_index;
+
+  // Buffer used by this endpoint
+  const uint32_t buffer = mmio_region_read_mask32(
+      usbdev->base_addr, config_in_reg_offset, USBDEV_CONFIGIN0_BUFFER0_MASK,
+      USBDEV_CONFIGIN0_BUFFER0_OFFSET);
+
+  // Check the status of the packet
+  if (mmio_region_get_bit32(usbdev->base_addr, config_in_reg_offset,
+                            USBDEV_CONFIGIN0_RDY0)) {
+    // Packet is marked as ready to be sent and pending transmission
+    *status = kDifUsbdevTxPacketStatusPending;
+  } else if (mmio_region_get_bit32(usbdev->base_addr, USBDEV_IN_SENT_REG_OFFSET,
+                                   endpoint_bit_index)) {
+    // Packet was sent successfully
+    // Clear IN_SENT bit (rw1c)
+    mmio_region_write32(usbdev->base_addr, USBDEV_IN_SENT_REG_OFFSET,
+                        1 << endpoint_bit_index);
+    // Return the buffer back to the free buffer pool
+    buffer_pool_add(&usbdev->buffer_pool, buffer);
+    *status = kDifUsbdevTxPacketStatusSent;
+  } else if (mmio_region_get_bit32(usbdev->base_addr, config_in_reg_offset,
+                                   USBDEV_CONFIGIN0_PEND0)) {
+    // Canceled due to an IN SETUP packet
+    // Clear pending bit (rw1c)
+    mmio_region_write32(usbdev->base_addr, config_in_reg_offset,
+                        1 << USBDEV_CONFIGIN0_PEND0);
```

Comment:
This can be a write-only set bit (your PR to fix my linker snafu got in).

(This applies elsewhere in the file, ofc.)